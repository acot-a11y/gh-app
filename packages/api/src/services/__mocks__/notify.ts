import type { MockFactory } from '../../__mocks__/types';
import type { NotifyService } from '../notify';

export const createNotifyServiceMock: MockFactory<NotifyService> = (
  mocks = {},
) => ({
  comment: jest.fn(),
  ...mocks,
});
