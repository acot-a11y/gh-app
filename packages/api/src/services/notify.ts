import type { PrCommentMeta } from '@acot/gh-app-types';
import type { Summary } from '@acot/types';
import { Align, getMarkdownTable } from 'markdown-table-ts';
import ms from 'pretty-ms';
import { GH_APP_LOGIN } from '../constants';
import type { GhIssueComment } from '../drivers/github';
import type { DriverRegistry } from '../drivers/registry';
import type { Logger } from '../logging';
import type { ServiceOutput } from './utils';
import { output } from './utils';

export type NotifyCommentInput = {
  token: string;
  owner: string;
  repo: string;
  number: number;
  meta: PrCommentMeta;
  summary: Summary;
};

export type NotifyCommentOutput = ServiceOutput<
  GhIssueComment,
  {
    error: string;
  }
>;

export type NotifyService = {
  comment: (input: NotifyCommentInput) => Promise<NotifyCommentOutput>;
};

const formatMessage = (message: string) => {
  return message.trim().replace(/(?:\r\n|\r|\n)/g, '<br />');
};

const format = (meta: PrCommentMeta, summary: Summary) => {
  const symbols = {
    success: ':white_check_mark:',
    error: ':x:',
    warning: ':warning:',
  };

  const passCount = (n: number) => (n > 0 ? `${n}` : '-');
  const errorCount = (n: number) => (n > 0 ? `${n}` : '-');
  const warningCount = (n: number) => (n > 0 ? `${n}` : '-');

  // Environment
  let body = `\n## Run details

${getMarkdownTable({
  table: {
    head: ['', ''],
    body: [
      ['**Core**', `\`v${meta.core.version}\``],
      ['**Runner**', `\`${meta.runner.name} v${meta.runner.version}\``],
      ['**Origin**', meta.origin],
      ['**Duration**', `${ms(summary.duration)}`],
      ['**Commit**', meta.commit],
    ],
  },
  alignment: [Align.Left, Align.Left, Align.Left],
})}
`;

  // Summary
  body += `## Audit summary

${getMarkdownTable({
  table: {
    head: ['', 'Rule', 'Duration', 'Pass', 'Error', 'Warning'],
    body: [
      ...Object.entries(summary.rules).map(([rule, result]) => {
        let status = symbols.success;
        if (result.errorCount > 0) {
          status = symbols.error;
        } else if (result.warningCount > 0) {
          status = symbols.warning;
        }

        return [
          status,
          `\`${rule}\``,
          ms(result.duration),
          passCount(result.passCount),
          errorCount(result.errorCount),
          warningCount(result.warningCount),
        ];
      }),
      [
        '',
        '',
        ms(summary.duration),
        passCount(summary.passCount),
        errorCount(summary.errorCount),
        warningCount(summary.warningCount),
      ],
    ],
  },
  alignment: [
    Align.Center,
    Align.Left,
    Align.Right,
    Align.Right,
    Align.Right,
    Align.Right,
  ],
})}
`;

  // Details
  body += '\n## Audit details\n';
  if (summary.results.length === 0) {
    body += `:tada: All checks pass :tada:`;
  } else {
    const lines = [] as string[];

    summary.results.forEach((result) => {
      if (result.results.length === 0) {
        return;
      }

      if (result.errorCount === 0 && result.warningCount === 0) {
        return;
      }

      const short = [
        `${symbols.success} ${result.passCount}`,
        `${symbols.error} ${result.errorCount}`,
        `${symbols.warning} ${result.warningCount}`,
      ].join(' ');

      lines.push(`<details>
<summary>${result.url} - ${short}</summary>

${getMarkdownTable({
  table: {
    head: ['', 'Rule', 'Details'],
    body: result.results.reduce<string[][]>((acc, cur) => {
      if (cur.status === 'pass') {
        return acc;
      }

      const body = [formatMessage(`${cur.message}`)];

      if (cur.selector || cur.help) {
        body.push(`<br />`);
      }
      if (cur.selector) {
        body.push(`<br />at \`${cur.selector}\``);
      }
      if (cur.help) {
        body.push(`<br />see ${formatMessage(cur.help)}`);
      }

      acc.push([
        cur.status === 'error' ? symbols.error : symbols.warning,
        `\`${cur.rule}\``,
        body.join(''),
      ]);

      return acc;
    }, []),
  },
  alignment: [Align.Center, Align.Left, Align.Left],
})}
</details>`);
    });

    body += lines.join('\n');
  }

  // Footer
  body += `<p align="right">Generated by :heart: <a href="https://github.com/acot-a11y/acot">acot</a>.</p>`;

  return body;
};

export const createNotifyService = (
  logger: Logger,
  driver: DriverRegistry,
): NotifyService => {
  const github = driver.resolve('github');

  return {
    comment: async ({ token, owner, repo, number, meta, summary }) => {
      const body = format(meta, summary);

      try {
        const comments = await github.getIssueCommentList({
          token,
          owner,
          repo,
          issue_number: number,
          per_page: 100,
          page: 1,
        });

        const found = comments.findIndex(
          ({ user }) => user.type === 'Bot' && user.login === GH_APP_LOGIN,
        );

        logger.debug({ found }, 'comments search index');

        const comment =
          found < 0
            ? await github.createIssueComment({
                token,
                owner,
                repo,
                issue_number: number,
                body,
              })
            : await github.updateIssueComment({
                token,
                owner,
                repo,
                comment_id: comments[found].id,
                body,
              });

        logger.debug(
          { comment },
          `${found < 0 ? 'created' : 'updated'} comment`,
        );

        return output(true, comment);
      } catch (e) {
        logger.error(e, 'comment failure');
        return output(false, {
          error: e instanceof Error ? e.message : '',
        });
      }
    },
  };
};

createNotifyService.inject = ['logger', 'driver'] as const;
