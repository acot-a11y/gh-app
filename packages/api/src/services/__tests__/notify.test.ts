import { createSummary } from '@acot/factory';
import type { PrCommentMeta } from '@acot/gh-app-types';
import type { Summary } from '@acot/types';
import type { PartialDeep } from 'type-fest';
import { format } from '../notify';

const createPrCommentMeta = (
  props: PartialDeep<PrCommentMeta> = {},
): PrCommentMeta => ({
  commit: 'abc12345',
  origin: 'http://localhost:3000',
  ...props,
  core: {
    version: '0.0.0',
    ...(props?.core ?? {}),
  },
  runner: {
    name: 'test',
    version: '0.0.0',
    ...(props?.core ?? {}),
  },
});

describe('NotifyService', () => {
  describe('internal#format', () => {
    test.each<[string, PrCommentMeta, Summary]>([
      ['empty', createPrCommentMeta(), createSummary()],
      [
        'mixed',
        createPrCommentMeta(),
        createSummary({
          rules: {
            '@acot/test/rule1': {
              duration: 1,
              passCount: 2,
              warningCount: 3,
              errorCount: 4,
            },
            '@acot/test/rule2': {
              duration: 2,
              passCount: 1,
            },
          },
          results: [
            {
              url: 'http://localhost:3000/',
              passCount: 2,
              warningCount: 3,
              errorCount: 2,
              rules: {
                '@acot/test/rule1': {
                  passCount: 1,
                  warningCount: 3,
                },
                '@acot/test/rule2': {
                  passCount: 1,
                },
              },
              results: [
                {
                  status: 'pass',
                  rule: '@acot/test/rule1',
                },
                {
                  status: 'warn',
                  rule: '@acot/test/rule1',
                  message: 'group1',
                },
                {
                  status: 'warn',
                  rule: '@acot/test/rule1',
                  message: 'group1',
                },
                {
                  status: 'warn',
                  rule: '@acot/test/rule1',
                  message: 'group1',
                  selector: '#rule1-warn3',
                  help: 'https://acot.example/help',
                },
                {
                  status: 'pass',
                  rule: '@acot/test/rule2',
                },
              ],
            },
            {
              url: 'http://localhost:3000/path/to',
              errorCount: 4,
              rules: {
                '@acot/test/rule1': {
                  errorCount: 4,
                },
              },
              results: [
                {
                  status: 'error',
                  rule: '@acot/test/rule1',
                  message: 'group1',
                  selector: '#rule1-error1',
                },
                {
                  status: 'error',
                  rule: '@acot/test/rule1',
                  message: 'group1',
                  selector: '#rule1-error2',
                },
                {
                  status: 'error',
                  rule: '@acot/test/rule1',
                  message: 'group2',
                  help: 'https://acot.example/help',
                },
                {
                  status: 'error',
                  rule: '@acot/test/rule1',
                  message: 'group2',
                  help: 'https://acot.example/help',
                },
              ],
            },
            {
              url: 'http://localhost:3000/empty',
              rules: {},
              results: [],
            },
          ],
        }),
      ],
    ])('%s', (_, meta, summary) => {
      expect(format(meta, summary)).toMatchSnapshot();
    });
  });
});
