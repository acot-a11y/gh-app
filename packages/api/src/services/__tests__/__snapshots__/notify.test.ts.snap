// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotifyService internal#format empty 1`] = `
"
## Run details

|  |  |
| :-- | :-- |
| **Core** | \`v0.0.0\` |
| **Runner** | \`test v0.0.0\` |
| **Origin** | http://localhost:3000 |
| **URLs** | 0 |
| **Duration** | 0ms |
| **Commit** | abc12345 |

## Audit details

:tada: All checks pass :tada:<p align=\\"right\\">Generated by :heart: <a href=\\"https://github.com/acot-a11y/acot\\">acot</a>.</p>"
`;

exports[`NotifyService internal#format mixed 1`] = `
"
## Run details

|  |  |
| :-- | :-- |
| **Core** | \`v0.0.0\` |
| **Runner** | \`test v0.0.0\` |
| **Origin** | http://localhost:3000 |
| **URLs** | 6 |
| **Duration** | 0ms |
| **Commit** | abc12345 |

## Audit summary

|  | Rule | Duration | Pass | Error | Warning |
| :-: | :-- | --: | --: | --: | --: |
| :x: | \`@acot/test/rule1\` | 1ms | 2 | 4 | 3 |
| :white_check_mark: | \`@acot/test/rule2\` | 2ms | 1 | - | - |
|  |  |  | - | - | - |

## Audit details

<details>
<summary><a href=\\"http://localhost:3000/\\">/</a> - :white_check_mark: 2 :warning: 3 :x: 2</summary>

|  | Rule | Details |
| :-: | :-- | :-- |
| :warning: | \`@acot/test/rule1\` | group1 |
| :warning: | \`@acot/test/rule1\` | group1<br /><br />at \`#rule1-warn3\`<br />see https://acot.example/help |
</details>

<details>
<summary><a href=\\"http://localhost:3000/path/to\\">/path/to</a> - :x: 4</summary>

|  | Rule | Details |
| :-: | :-- | :-- |
| :x: | \`@acot/test/rule1\` | group1<br /><br />at \`#rule1-error1\`<br />at \`#rule1-error2\` |
| :x: | \`@acot/test/rule1\` | group2<br /><br />see https://acot.example/help |
</details><p align=\\"right\\">Generated by :heart: <a href=\\"https://github.com/acot-a11y/acot\\">acot</a>.</p>"
`;
